doc/Authoring_Extensions.md

```diff
@@ -0,0 +1,250 @@
+# Authoring Theia Extensions
+
+As an example, we are going to add a menu item _Say hello_ that display a notification "Hello world!". This article is guiding you through all the necessary steps.
+
+## Theia’s Architecture
+
+A Theia app is composed of so called _extensions_. An extension provides a set of widgets, commands, handlers, etc. for a specific functionality. Theia itself ships a number of extensions, e.g. for editors, terminals, the project view etc. The infrastructure of adding/removing extensions dynamically is still under construction and not covered in this article.
+
+Theia defines a plethora of contribution interfaces that allow extensions to add their behaviour to various aspects of the application. Just search for interfaces with the name `*Contribution` to get an idea. An extension implements the contribution interfaces belonging to the functionality it wants to deliver. In this example, we are going to implement a `CommandContribution` and a `MenuContribution`. 
+
+In Theia, everything is wired up via [dependency injection](Architecture.md#dependency-injection-di). An extension defines one ore more dependency injection modules. This is where it binds its contribution implementations to the respective contribution interface. The modules are listed in the `package.json` of the extension package. When the application starts, the union of all these modules is used to configure a single, global dependency injection container. The runtime will then collect all contributions of a specific kind by means of a multi-inject.
+
+## Prerequisites
+
+You’ll need node in version 8:
+
+```bash
+curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.5/install.sh | bash
+nvm install 8
+```
+
+and yarn in version 1.0.2
+
+```
+npm install -g yarn@1.0.2
+```
+
+Also make sure your `python --version` points to a Python 2.x installation.
+
+## Project Layout
+
+We are going to create a monorepo (a repository containing multiple npm packages) named `theia-hello-world-extension` containing three packages: `hello-world-extension`, `browser-app` and `electron-app`. The first contains our extension, the latter two the Theia applications to run our extension in browser and electron mode.  We are going to use `yarn` instead of `npm`, because it allows to structure such monorepos into workspaces. In our case, each workspace contains its own `npm` package. Common dependencies of these packages are 'hoisted' by `yarn` to their common root directory. We are also going to use `lerna` to run scripts accross workspaces. 
+
+The setup of such a repository is a bit tricky. This is why we have created a [code generator](https://www.npmjs.com/package/generator-theia-extension) to scaffold the project. It will also generate the `hello-world` example. Run it using
```

Comment:
The 'tricky' sentence sounds a bit negative. I think we can just leave it out and and say that a generator makes this super easy.